contracts/engine-registry/CoreRegistryV1.sol

```diff
@@ -0,0 +1,240 @@
+ * contract being registered during a transaction originating from the
+ * owner:
+ * - registerContract
+ * ----------------------------------------------------------------------------
+ * The following function is restricted to the contract owner sending, or the
+ * the contract being unregistered:
+ * - unregisterContract
+ * ----------------------------------------------------------------------------
+ * The following functions are restricted to the contract owner:
+ * - registerContracts
+ * - unregisterContracts
+ * - Ownable: transferOwnership
+ * - Ownable: renounceOwnership
+ * ----------------------------------------------------------------------------
+ * Additional privileged roles may be described on minters, registries, and
+ * other contracts that may interact with this contract.
+ */
+contract CoreRegistryV0 is Ownable, ICoreRegistryV1 {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    /// private enumerable set of registered contracts
+    EnumerableSet.AddressSet private registeredContracts;
+
+    /// private mapping of registered contract addresses to
+    EnumerableSet.AddressSet
+        private registeredMinterFilterV2CompatibleContracts;
+
+    constructor() Ownable() {}
+
+    /**
+     * @notice Register a contract and emit a `ContractRegistered` event with
+     * the provided information. Only callable by the owner or the contract
+     * being registered, and only if tx.origin == owner.
+     * Reverts if authorization fails, or if the contract is already
+     * registered.
+     */
+    function registerContract(
+        address _contractAddress,
+        bytes32 _coreVersion,
+        bytes32 _coreType
+    ) external {
+        // CHECKS
+        // Validate against `tx.origin` rather than `msg.sender` as it is intended that this registration be
+        // performed in an automated fashion *at the time* of contract deployment for the `_contractAddress`.
+        require(tx.origin == owner(), "Only tx origin of owner");
+        // Prevent registration of a contract by an unrelated contract
+        require(
+            msg.sender == _contractAddress || msg.sender == owner(),
+            "Only calls by owner or contract"
+        );
+        // EFFECTS
+        _registerContract(_contractAddress, _coreVersion, _coreType);
+    }
+
+    /**
+     * @notice Unregister a contract and emit a `ContractUnregistered` event.
+     * Only callable by the owner or this contract or the contract being
+     * unregistered.
+     * Reverts if authorization fails, or if the contract is not already
+     * registered.
+     */
+    function unregisterContract(address _contractAddress) external {
+        // CHECKS
+        // Prevent unregistration of a contract by an unrelated contract
+        require(
+            msg.sender == _contractAddress || msg.sender == owner(),
+            "Only calls by owner or contract"
+        );
+        // EFFECTS
+        _unregisterContract(_contractAddress);
+    }
```

Comment:
@ryley-o I don't think we want to allow contracts to register/unregister _themselves_ 🤔

Reply:
hmm...

For registering, I think we want to allow a contract to register itself if `tx.origin == owner`, because that enables the deployment process where the contract register's itself via a call from the contract's constructor. `EngineRegistryV0` only checks that `tx.origin` is the deployer address (aka `owner()` in this update), which we still enforce. Adding an additional constraint on `msg.sender` adds a small amount of security benefit (albeit minor).

For unregistering, `EngineRegistryV0` only checks that `tx.origin` is the deployer address (aka `owner()`), which is not ideal from a security perspective. I think this version should definitely change that check to be `msg.sender == owner()`. I would be okay with not allowing contracts to unregister themselves (since there isn't really a reason I can think of why someone would fork the V3 Engine contract to do that).

Reply:
Hrm, I have some thoughts on why this may create somewhat of a problematic pattern – let's discuss more on Monday! 🙏

Reply:
TODO: I'll remove the ability for a contract to eject itself 🙏

Reply:
updated and resolving, based on today's conversation with @jakerockland 🙏